sys/include/net/nanocoap.h

```diff
@@ -250,7 +250,7 @@ typedef ssize_t (*coap_handler_t)(coap_pkt_t *pkt, uint8_t *buf, size_t len, voi
  */
 typedef struct {
     const char *path;               /**< URI path of resource               */
-    unsigned methods;               /**< OR'ed methods this resource allows */
+    uint32_t methods;               /**< OR'ed methods this resource allows */
```

Comment:
to me it definitely makes sense to have a specific size here, but I also agree with @kaspar030 and @maribu that `uint16_t` should be sufficient (for now).

Reply:
also it might be necessary to adapt some methods, like `coap_method2flag`?

Reply:
I don't recall having said that, but I do agree anyway ;-)

Reply:
oops, right ... was @kb2ma

Reply:
Ping @bergzand?

Reply:
For reference: [HTTP has 39 methods](https://www.iana.org/assignments/http-methods/http-methods.xhtml) ;-).

Reply:
so you think `uint8_t` would do?

Reply:
I think 16 or 32 to bits is a good margin. CoAP only has space for [32 methods](https://tools.ietf.org/html/rfc7252#section-3) I realize now.

Reply:
Just as a reminder: this is a bit field not a value, so we need |methods| bits, not values ;-)

Reply:
A solution tipsy @cladmi is screaming at me would also be to define a `coap_method_flags_t` and 
(and set it to `uint8_t` for now) and be super-extensible in the future (just change the type definition ;-) and maybe some macros).

Reply:
So far 7 methods are officially standardized, see [IANA entry](https://www.iana.org/assignments/core-parameters/core-parameters.xhtml#method-codes). (`GET`, `POST`, `PUT`, and `DELETE` are introduced by [RFC  7252](https://tools.ietf.org/html/rfc7252), `FETCH`, `PATCH` and `iPATCH` were added in [RFC 8132](https://tools.ietf.org/html/rfc8132).) So *for* *now* `uint8_t` would be enough. Keeping in mind that there is now way to have more than 31 methods, `uint32_t` would be completely future proof.

But honestly: CoAP is intended for constrained devices. The addition of new methods is hopefully done with care and only for very good reason. I personally think that `uint16_t` should be enough for quite some time.

Reply:
+1 for using an `enum`

Reply:
I didn't say `enum`. From my experience, using `enum`s for flags lead to insufferable pain. I meant something like

```C
typedef coap_method_flags_t uint8_t
```

;-)

Reply:
> Keeping in mind that there is no way to have more than 31 methods, `uint32_t` would be completely future proof.

Let's not forget that this is an internal field. It is fine to use the smallest type that fits *currently supported* flags. This is not a field that might suddenly be populated over the network with unexpected values.

Regarding the struct's size, alignment will make any change not really save space. A "smaller than register" value might add some masking code, though.

Honestly, the current "at least 16bit" which unsigned implies seems totally fine to me...

Reply:
> I didn't say `enum`. From my experience, using `enum`s for flags lead to insufferable pain.

Why not using an anonymous `enum` instead of `#define`s for the values? The bitmask can still be stored in `uint8_t`. I made good experience with this approach ;-)

+1 for using a `typedef` to allow future changes

Reply:
We should keep in mind that other options (not just methods) may arise in the future, as it is being proposed in #11098, so if we are planning on placing them under this field it might be worth changing it to `uint32_t`.

Reply:
I've changed the method to a `typedef` as suggested by @miri64 (or @cladmi) [here](https://github.com/RIOT-OS/RIOT/pull/10214#discussion_r260477918). No measurable difference in code size.